shark-hprof/src/main/java/shark/HprofDeobfuscator.kt

```diff
@@ -0,0 +1,237 @@
+            when (record) {
+              is StringRecord -> {
+                maxId = maxId.coerceAtLeast(record.id)
+                hprofStringCache[record.id] = record.string
+              }
+              is LoadClassRecord -> {
+                maxId = maxId.coerceAtLeast(record.id)
+                classNames[record.id] = record.classNameStringId
+              }
+              is StackFrameRecord -> maxId = maxId.coerceAtLeast(record.id)
+              is ObjectRecord -> {
+                maxId = when (record) {
+                  is ClassDumpRecord -> maxId.coerceAtLeast(record.id)
+                  is ClassSkipContentRecord -> maxId.coerceAtLeast(record.id)
+                  is InstanceDumpRecord -> maxId.coerceAtLeast(record.id)
+                  is InstanceSkipContentRecord -> maxId.coerceAtLeast(record.id)
+                  is ObjectArrayDumpRecord -> maxId.coerceAtLeast(record.id)
+                  is ObjectArraySkipContentRecord -> maxId.coerceAtLeast(record.id)
+                  is PrimitiveArrayDumpRecord -> maxId.coerceAtLeast(record.id)
+                  is PrimitiveArraySkipContentRecord -> maxId.coerceAtLeast(record.id)
+                }
+              }
+            }
+          })
+      }
+
+    return maxId
+  }
+
+  @Suppress("LongParameterList")
+  private fun writeHprofRecords(
+    inputHprofFile: File,
+    outputHprofFile: File,
+    proguardMapping: ProguardMapping,
+    hprofStringCache: MutableMap<Long, String>,
+    classNames: MutableMap<Long, Long>,
+    firstId: Long
+  ): File {
+    var id = firstId
+
+    Hprof.open(inputHprofFile)
+      .use { hprof ->
+        val reader = hprof.reader
+        HprofWriter.open(
+          outputHprofFile,
+          identifierByteSize = reader.identifierByteSize,
+          hprofVersion = hprof.hprofVersion
+        ).use { writer ->
+          reader.readHprofRecords(setOf(HprofRecord::class),
+            OnHprofRecordListener { _,
+              record ->
+              // HprofWriter automatically emits HeapDumpEndRecord, because it flushes
+              // all continuous heap dump sub records as one heap dump record.
+              if (record is HeapDumpEndRecord) {
+                return@OnHprofRecordListener
+              }
+
+              when (record) {
+                is StringRecord -> {
+                  writer.write(
+                    createDeobfuscatedStringRecord(record, proguardMapping, hprofStringCache)
+                  )
+                }
+                is HeapDumpRecord -> {
+                  when (record) {
+                    is ObjectRecord -> {
+                      when (record) {
+                        is ClassDumpRecord -> {
+                          val recordsToWrite = createDeobfuscatedClassDumpRecord(
+                            record, proguardMapping, hprofStringCache, classNames, id
+                          )
+                          id += recordsToWrite.size
```

Comment:
Might want to be a bit less aggressive if we stop creating new records for non deobfuscated fields (e.g. have `createDeobfuscatedClassDumpRecord` return the new max id)

Reply:
Not sure if I understood it correctly but I changed it